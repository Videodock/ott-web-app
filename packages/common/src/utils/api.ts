import type { CommonResponse } from '@inplayer-org/inplayer.js';

import type { InPlayerError } from '../../types/inplayer';

export class ApiError extends Error {
  code: number;
  message: string;

  constructor(message = '', code: number) {
    super(message);
    this.name = 'ApiError';
    this.message = message;
    this.code = code;
  }
}

/**
 * Get data
 * @param response
 */
export const getDataOrThrow = async (response: Response) => {
  const data = await response.json();

  if (!response.ok) {
    const message = `Request '${response.url}' failed with ${response.status}`;
    const apiMessage = data && typeof data === 'object' && 'message' in data && typeof data.message === 'string' ? data.message : undefined;

    throw new ApiError(apiMessage || message, response.status || 500);
  }

  return data;
};

export const getCommonResponseData = (response: { data: CommonResponse }) => {
  const { code, message } = response.data;
  if (code !== 200) {
    throw new Error(message);
  }
  return {
    errors: [],
    responseData: {
      message,
      code,
    },
  };
};

export const isCommonError = (error: unknown): error is InPlayerError => {
  return (
    typeof error === 'object' &&
    error !== null &&
    'response' in error &&
    typeof (error as InPlayerError).response?.data?.code === 'number' &&
    typeof (error as InPlayerError).response?.data?.message === 'string'
  );
};
