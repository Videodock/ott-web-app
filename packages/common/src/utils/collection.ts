import type { CustomFormField, ConsentsValue } from '../../types/account';
import type { Config } from '../../types/config';
import type { GenericFormValues } from '../../types/form';
import type { Playlist, PlaylistItem } from '../../types/playlist';
import { CARD_ASPECT_RATIOS } from '../constants';

export type PosterAspectRatio = (typeof CARD_ASPECT_RATIOS)[number];

const getFiltersFromConfig = (config: Config, playlistId: string | undefined): string[] => {
  const menuItem = config.menu.find((item) => item.contentId === playlistId);
  const filters = menuItem?.filterTags?.split(',').filter(Boolean);

  return filters || [];
};

const filterPlaylist = (playlist: Playlist, filter: string) => {
  if (filter === '') return playlist;

  return {
    ...playlist,
    playlist: playlist.playlist.filter(({ tags }) => (tags ? tags.split(',').includes(filter) : false)),
  };
};

const chunk = <T>(input: T[], size: number) => {
  return input?.reduce((arr: T[][], item, idx: number) => {
    return idx % size === 0 ? [...arr, [item]] : [...arr.slice(0, -1), [...arr.slice(-1)[0], item]];
  }, []);
};

const findPlaylistImageForWidth = (playlistItem: PlaylistItem, width: number): string =>
  playlistItem.images.find((img) => img.width === width)?.src || playlistItem.image;

const generatePlaylistPlaceholder = (playlistLength: number = 15): Playlist => ({
  title: '',
  playlist: new Array(playlistLength).fill({}).map(
    (_value, index) =>
      ({
        description: '',
        duration: 0,
        feedid: '',
        image: '',
        images: [],
        cardImage: '',
        backgroundImage: '',
        channelLogoImage: '',
        link: '',
        genre: '',
        mediaid: `placeholder_${index}`,
        pubdate: 0,
        rating: '',
        sources: [],
        tags: '',
        title: '',
        tracks: [],
      } as PlaylistItem),
  ),
});

const formatConsentValues = (publisherConsents: CustomFormField[] | null = [], customerConsents: ConsentsValue[] | null = []) => {
  if (!publisherConsents || !customerConsents) {
    return {};
  }

  return publisherConsents.reduce((acc, publisherConsent) => {
    const consent = customerConsents?.find((customerConsent) => customerConsent.name === publisherConsent.name);

    if (consent) {
      acc[publisherConsent.name] = consent.state === 'accepted';
    }

    return acc;
  }, {} as Record<string, boolean>);
};

const formatConsents = (publisherConsents: CustomFormField[] | null = [], customerConsents: ConsentsValue[] | null = []) => {
  if (!publisherConsents || !customerConsents) {
    return {};
  }
  const values: Record<string, boolean> = {};
  publisherConsents?.forEach((publisherConsent) => {
    if (customerConsents?.find((customerConsent) => customerConsent.name === publisherConsent.name && customerConsent.state === 'accepted')) {
      values[publisherConsent.name] = true;
    }
  });

  return values;
};

const extractConsentValues = (consents?: CustomFormField[]) => {
  const values: Record<string, string | boolean> = {};

  if (!consents) {
    return values;
  }

  consents?.forEach((consent) => {
    values[consent.name] = consent.type === 'checkbox' ? consent.enabledByDefault === true : consent.defaultValue ?? '';
  });

  return values;
};

const formatConsentsFromValues = (publisherConsents: CustomFormField[] | null, values?: GenericFormValues) => {
  const consents: ConsentsValue[] = [];

  if (!publisherConsents || !values) return consents;

  publisherConsents.forEach((consent) => {
    consents.push({
      name: consent.name,
      version: consent.version || '1',
      state: values[consent.name] ? 'accepted' : 'declined',
    });
  });

  return consents;
};

const checkConsentsFromValues = (publisherConsents: CustomFormField[], consents: Record<string, string | boolean>) => {
  const customerConsents: ConsentsValue[] = [];
  const consentsErrors: string[] = [];

  if (!publisherConsents || !consents) return { customerConsents, consentsErrors };

  publisherConsents.forEach((consent) => {
    if (consent.required && !consents[consent.name]) {
      consentsErrors.push(consent.name);
    }

    customerConsents.push({
      name: consent.name,
      version: consent.version || '1',
      state: consents[consent.name] ? 'accepted' : 'declined',
    });
  });

  return { customerConsents, consentsErrors };
};

const deepCopy = (obj: unknown) => {
  if (Array.isArray(obj) || (typeof obj === 'object' && obj !== null)) {
    return JSON.parse(JSON.stringify(obj));
  }
  return obj;
};

const parseAspectRatio = (input: unknown) => {
  if (typeof input === 'string' && (CARD_ASPECT_RATIOS as readonly string[]).includes(input)) return input as PosterAspectRatio;
};

const parseTilesDelta = (posterAspect?: PosterAspectRatio) => {
  if (!posterAspect) {
    return 0;
  }

  const parts = posterAspect.split(':');

  return parts.length === 2 ? Math.floor(parseInt(parts[1]) / parseInt(parts[0])) : 0;
};

export {
  getFiltersFromConfig,
  filterPlaylist,
  chunk,
  findPlaylistImageForWidth,
  generatePlaylistPlaceholder,
  formatConsentValues,
  formatConsents,
  formatConsentsFromValues,
  extractConsentValues,
  checkConsentsFromValues,
  deepCopy,
  parseAspectRatio,
  parseTilesDelta,
};
