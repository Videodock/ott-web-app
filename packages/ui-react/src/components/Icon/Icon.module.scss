@use '@jwp/ott-ui-react/src/styles/variables';

.icon {
  width: 24px;
  height: 24px;
  fill: currentColor;
}

.profileIcon {
  width: 24px;
  height: 24px;
  border-radius: 50%;
}
