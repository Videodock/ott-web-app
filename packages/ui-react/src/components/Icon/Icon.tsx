import classNames from 'classnames';
import React, { FC, SVGProps } from 'react';

import styles from './Icon.module.scss';

export type Props = {
  icon: FC<SVGProps<SVGSVGElement>>;
  className?: string;
};

const Icon: FC<Props> = ({ icon: IconComponent, className }) => {
  return <IconComponent className={classNames(styles.icon, className)} aria-hidden="true" />;
};

export default Icon;
