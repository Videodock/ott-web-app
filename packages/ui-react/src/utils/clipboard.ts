export const copyToClipboard = (value: string): void => {
  const inputElement = document.createElement('input');
  inputElement.style.zIndex = '-10';
  inputElement.style.position = 'absolute';
  inputElement.style.left = '-1000';
  inputElement.value = value;
  document.body.appendChild(inputElement);
  inputElement.select();
  document.execCommand('copy');
  inputElement.blur();
  document.body.removeChild(inputElement);
};
