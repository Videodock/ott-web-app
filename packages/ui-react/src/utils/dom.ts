export const canUseDOM = !!(typeof window !== 'undefined' && window.document && window.document.createElement);

let size: number;

export default function scrollbarSize(recalc?: boolean) {
  if ((!size && size !== 0) || recalc) {
    if (canUseDOM) {
      const scrollDiv = document.createElement('div');
      scrollDiv.style.position = 'absolute';
      scrollDiv.style.top = '-9999px';
      scrollDiv.style.width = '50px';
      scrollDiv.style.height = '50px';
      scrollDiv.style.overflow = 'scroll';
      document.body.appendChild(scrollDiv);
      size = scrollDiv.offsetWidth - scrollDiv.clientWidth;
      document.body.removeChild(scrollDiv);
    }
  }
  return size;
}

export const addScript = (src: string): Promise<void> => {
  return new Promise((resolve) => {
    const script: HTMLScriptElement = document.createElement('script');
    script.type = 'text/javascript';
    script.async = true;
    script.src = src;
    script.onload = () => resolve();
    script.onerror = (error) => {
      console.info('Error loading external script', error);
      throw new Error('Error loading external script');
    };
    document.body.appendChild(script);
  });
};
