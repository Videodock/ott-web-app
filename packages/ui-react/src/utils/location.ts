import { createURL, type QueryParamsArg } from '@jwp/ott-common/src/utils/urlFormatting';
import type { Location } from 'react-router-dom';

import type { AccountModals } from '../containers/AccountModal/AccountModal';

export const createURLFromLocation = (location: Location, queryParams: QueryParamsArg = {}) => {
  return createURL(`${location.pathname}${location.search || ''}`, queryParams);
};

export const modalURLFromLocation = (location: Location, u: keyof AccountModals | null, queryParams?: QueryParamsArg) => {
  return createURL(`${location.pathname}${location.search || ''}`, { u, ...queryParams });
};
