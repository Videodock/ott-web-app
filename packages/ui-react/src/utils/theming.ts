import type { Config } from '@jwp/ott-common/types/config';
import { calculateContrastColor } from '@jwp/ott-common/src/utils/common';
import env from '@jwp/ott-common/src/env';

export const setThemingVariables = (config: Config) => {
  const root = document.querySelector(':root') as HTMLElement;
  const { backgroundColor, highlightColor, headerBackground } = config.styling || {};
  const bodyFont = env.APP_BODY_FONT;
  const bodyAltFont = env.APP_BODY_ALT_FONT;

  if (!root) return;

  if (backgroundColor) {
    root.style.setProperty('--body-background-color', backgroundColor);
    root.style.setProperty('--background-contrast-color', calculateContrastColor(backgroundColor));
  }
  if (highlightColor) {
    root.style.setProperty('--highlight-color', highlightColor);
    root.style.setProperty('--highlight-contrast-color', calculateContrastColor(highlightColor));
  }
  if (headerBackground) {
    root.style.setProperty('--header-background', headerBackground);
    root.style.setProperty('--header-contrast-color', calculateContrastColor(headerBackground));
  }

  if (bodyFont) {
    root.style.setProperty('--body-font-family', bodyFont);
  }
  if (bodyAltFont) {
    root.style.setProperty('--body-alt-font-family', bodyAltFont);
  }
};
