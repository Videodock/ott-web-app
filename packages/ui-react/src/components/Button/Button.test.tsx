import React from 'react';
import { axe } from 'vitest-axe';
import AccountCircle from '@jwp/ott-theme/assets/icons/account_circle.svg?react';
import 'vitest-axe/extend-expect';

import { renderWithRouter } from '../../../test/utils';
import Icon from '../Icon/Icon';

import Button from './Button';

describe('<Button>', () => {
  test('renders and matches snapshot', () => {
    const { container } = renderWithRouter(
      <>
        <Button label="Default button" />
        <Button label="Active button" active />
        <Button label="Disabled button" disabled />
        <Button label="Busy button" busy />
        <Button label="Small button" size="small" />
        <Button label="Large button" size="large" />
        <Button label="Full width button" fullWidth />
        <Button label="With start icon" startIcon={<Icon icon={AccountCircle} />} />
        <Button label="Button link" to="https://jwplayer.com" />
      </>,
    );

    expect(container).toMatchSnapshot();
  });

  test('WCAG 2.2 (AA) compliant', async () => {
    const { container } = renderWithRouter(
      <>
        <Button label="Default button" />
        <Button label="Active button" active />
        <Button label="Disabled button" disabled />
        <Button label="Busy button" busy />
        <Button label="Small button" size="small" />
        <Button label="Large button" size="large" />
        <Button label="Full width button" fullWidth />
        <Button label="With start icon" startIcon={<Icon icon={AccountCircle} />} />
        <Button label="Button link" to="https://jwplayer.com" fullWidth />
      </>,
    );

    expect(await axe(container, { runOnly: ['wcag21a', 'wcag21aa', 'wcag22aa'] })).toHaveNoViolations();
  });
});
