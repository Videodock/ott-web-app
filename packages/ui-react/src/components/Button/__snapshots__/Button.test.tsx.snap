// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Button> > renders and matches snapshot 1`] = `
<div>
  <button
    class="_button_f8f296 _default_f8f296 _outlined_f8f296"
    type="button"
  >
    <span>
      Default button
    </span>
  </button>
  <button
    class="_button_f8f296 _default_f8f296 _outlined_f8f296 _active_f8f296 "
    type="button"
  >
    <span>
      Active button
    </span>
  </button>
  <button
    aria-disabled="true"
    class="_button_f8f296 _default_f8f296 _outlined_f8f296 _disabled_f8f296"
    disabled=""
    type="button"
  >
    <span>
      Disabled button
    </span>
  </button>
  <button
    class="_button_f8f296 _default_f8f296 _outlined_f8f296"
    type="button"
  >
    <span
      class="_hidden_f8f296"
    >
      Busy button
    </span>
    <div
      class="_buffer_d122df _centerAbsolute_f8f296 _small_d122df"
    >
      <div />
      <div />
      <div />
      <div />
    </div>
  </button>
  <button
    class="_button_f8f296 _default_f8f296 _outlined_f8f296 _small_f8f296"
    type="button"
  >
    <span>
      Small button
    </span>
  </button>
  <button
    class="_button_f8f296 _default_f8f296 _outlined_f8f296 _large_f8f296"
    type="button"
  >
    <span>
      Large button
    </span>
  </button>
  <button
    class="_button_f8f296 _default_f8f296 _outlined_f8f296 _fullWidth_f8f296"
    type="button"
  >
    <span>
      Full width button
    </span>
  </button>
  <button
    class="_button_f8f296 _default_f8f296 _outlined_f8f296"
    type="button"
  >
    <div
      class="_startIcon_f8f296"
    >
      <svg
        aria-hidden="true"
        class="_icon_585b29"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12,19.2C9.5,19.2 7.29,17.92 6,16C6.03,14 10,12.9 12,12.9C14,12.9 17.97,14 18,16C16.71,17.92 14.5,19.2 12,19.2M12,5A3,3 0 0,1 15,8A3,3 0 0,1 12,11A3,3 0 0,1 9,8A3,3 0 0,1 12,5M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12C22,6.47 17.5,2 12,2Z"
        />
      </svg>
    </div>
    <span>
      With start icon
    </span>
  </button>
  <a
    class="_button_f8f296 _default_f8f296 _outlined_f8f296"
    href="https://jwplayer.com"
  >
    <span>
      Button link
    </span>
  </a>
</div>
`;
